### PR TITLE
Boyce index code

### PR DESCRIPTION
**Pull request by PC-FSU**
#### Summary:
In this pull request, I have added functionality to calculate the **continuous Boyce index** as described in [Hirzel et al. (2006)](https://www.whoi.edu/cms/files/hirzel_etal_2006_53457.pdf) . This method provides a reliable way to evaluate habitat suitability models, specifically for presence-only data. Along with the implementation, I have also added test cases to ensure the correctness and robustness of the new function.

#### Key Updates:
1. **Boyce Index Calculation**:
    - Implemented a function to compute the continuous Boyce index based on intervals of habitat suitability values.
    - This includes handling both sliding window and fixed-bin approaches, as well as ensuring compatibility with NumPy, Pandas, and GeoPandas data structures.
    - Added logic to handle edge cases like empty arrays, NaN values, and invalid inputs (e.g., non-1D arrays).

2. **Test Cases**:
    - Created test cases to validate the behavior of the Boyce index calculation under various scenarios (e.g., different bin sizes, presence of NaN values).
    - Ensured that the function calculates the correct predicted-to-expected (P/E) ratio and returns accurate Spearman correlation coefficients.

3. **Notebook Update**:
    - Updated the notebook `WorkingWithGeospatialData.ipynb` to include a detailed example demonstrating how to use the continuous Boyce index function.

#### Testing:
The test cases ensure that the continuous Boyce index function works as expected. The test cases cover:
- Different binning strategies.
- Varying habitat suitability ranges.
- Handling of NaN values in input data.

This PR enhances the project by providing a robust and well-tested method to evaluate habitat suitability models using presence-only data, with clear examples in the updated notebook.

### CODE REVIEW

File: elapid/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+from scipy.stats import spearmanr
+
+
+# implement Boyce index as describe in https://www.whoi.edu/cms/files/hirzel_etal_2006_53457.pdf (Eq.4)
+
+
+def boycei(interval, obs, fit):
```
Comment (earth-chris) [inline elapid/evaluate.py]: i might prefer renaming these functions `boyce_index` and `continuous_boyce_index` to differentiate (boycei/boyce_index are easily confused)

File: elapid/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    """
+    Calculate the Boyce index for a given interval.
+    
+    Args:
+        interval (tuple or list): Two elements representing the lower and upper bounds of the interval.
+        obs (numpy.ndarray): Observed suitability values (i.e., predictions at presence points).
+        fit (numpy.ndarray): Suitability values (e.g., from a raster), i.e., predictions at presence + background points.
```
Comment (earth-chris) [inline elapid/evaluate.py]: to better align with `sklearn` api design, prefer renaming variables and order them as `boyce_index(yobs, ypred, interval)`

File: elapid/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    if ei == 0:
+        fi = np.nan  # Avoid division by zero
+    else:
+        fi = pi / ei
+
+    return fi
+
+
+def boyce_index(fit, obs, nclass=0, window="default", res=100, PEplot=False):
```
Comment (earth-chris) [inline elapid/evaluate.py]: please rename and reorder fit, obs as `yobs, ypred`

File: elapid/evaluate.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    # Remove NaNs from fit and obs
+    fit = fit[~np.isnan(fit)]
+    obs = obs[~np.isnan(obs)]
+
+    if len(fit) == 0 or len(obs) == 0:
+        raise ValueError("After removing NaNs, 'fit' or 'obs' arrays cannot be empty.")
+
+    
+    # Remove NaNs from fit
+    fit = fit[~np.isnan(fit)]
```
Comment (earth-chris) [inline elapid/evaluate.py]: remove duplicate code

File: elapid/evaluate.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        vec_mov = np.linspace(mini, maxi - window, num=res+1)
+        intervals = np.column_stack((vec_mov, vec_mov + window))
+    elif isinstance(nclass, (list, np.ndarray)) and len(nclass) > 1:
+        nclass.sort()
+        if mini > nclass[0] or  maxi < nclass[-1]:
+            raise ValueError(f"The range provided via nclass is: ({nclass[0], nclass[-1]}). The range computed via fit is: ({mini, maxi}). Provided range via nclass should be in range computed via (max(fit), min(fit)).")
+        vec_mov = np.concatenate(([mini], nclass))
+        intervals = np.column_stack((vec_mov[:-1], vec_mov[1:]))
+        print(vec_mov)
+        print(intervals)
```
Comment (earth-chris) [inline elapid/evaluate.py]: remove debug print statements

File: elapid/evaluate.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    
+    # Remove NaNs from fit
+    fit = fit[~np.isnan(fit)]
+
+    if window == "default":
+        window = (np.max(fit) - np.min(fit)) / 10.0
+
+    mini = np.min(fit)
+    maxi = np.max(fit)
```
Comment (earth-chris) [inline elapid/evaluate.py]: cleaner to compute these above the window test then reuse the variables instead of multiple invocations

File: elapid/evaluate.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            raise ValueError(f"The range provided via nclass is: ({nclass[0], nclass[-1]}). The range computed via fit is: ({mini, maxi}). Provided range via nclass should be in range computed via (max(fit), min(fit)).")
+        vec_mov = np.concatenate(([mini], nclass))
+        intervals = np.column_stack((vec_mov[:-1], vec_mov[1:]))
+        print(vec_mov)
+        print(intervals)
+    elif nclass > 0:
+        vec_mov = np.linspace(mini, maxi, num=nclass + 1)
+        intervals = np.column_stack((vec_mov[:-1], vec_mov[1:]))
+    else:
+        raise ValueError("Invalid nclass value.")
```
Comment (earth-chris) [inline elapid/evaluate.py]: a comment or two in this section would be useful to elucidate the different methods for computing the intervals

File: elapid/evaluate.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    # use interval midpoints to calculate the spearmanr coeff. 
+    intervals_mid = np.mean(intervals[valid], axis=1)
+    if np.sum(valid) <= 2:
+        corr = np.nan
+    else:
+        f_valid = f[valid]
+        corr, _ = spearmanr(f_valid, intervals_mid)
+
+
+    if PEplot:
```
Comment (earth-chris) [inline elapid/evaluate.py]: I would prefer to keep plotting as a separate function, which makes for cleaner code.

File: elapid/evaluate.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    # Apply boycei function to each interval
+    f_list = []
+    for inter in intervals:
+        fi = boycei(inter, obs, fit)
+        f_list.append(fi)
+    f = np.array(f_list)
+
+
+    # Remove NaNs
+    valid = ~np.isnan(f)
```
Comment (earth-chris) [inline elapid/evaluate.py]: there seems to be a lot of `nan` checking. if the nans are removed from the initial arrays, what would lead to 'invalid' values? are we sure this isn't covering up some other issue in the calculations?

File: elapid/evaluate.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        plt.plot(intervals_mid, f[valid], marker='o')
+        plt.xlabel('Habitat suitability')
+        plt.ylabel('Predicted/Expected ratio')
+        plt.title('Boyce Index')
+        plt.show()
+
+
+    results = {
+        'F.ratio': f,
+        'Spearman.cor': round(corr, 3) if not np.isnan(corr) else np.nan,
```
Comment (earth-chris) [inline elapid/evaluate.py]: another suspicious nan check here. also, it seems unnecessary to round the correlation coefficient.

File: elapid/evaluate.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+            - 'F.ratio' (numpy.ndarray): The P/E ratio for each bin.
+            - 'Spearman.cor' (float): The Spearman's rank correlation coefficient between interval midpoints and F ratios.
+            - 'HS' (numpy.ndarray): The habitat suitability intervals.
+    
+    Example:
+        # Predicted suitability scores (e.g., predictions at presence + background points)
+        predicted = np.array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0])
+
+        # Observed presence suitability scores (e.g., predictions at presence points)
+        observed = np.array([0.3, 0.7, 0.8, 0.9])
```
Comment (earth-chris) [inline elapid/evaluate.py]: based on this example, it's not clear to me why users would want to pass in arrays of different lengths (where predicted and observed are not matching: one is presence+background, the other is just presence).

**Review by earth-chris — CHANGES_REQUESTED** [review]
Thank you for submitting this PR, @PC-FSU. Including this metric is a nice addition to the package. I have a few comments and requests.

- Thank you very much for the detailed docstring and jupyter examples.
- Thank you very very much for including tests.
- Please apply proper code formatting. To do this, review the [contributing guidelines](https://earth-chris.github.io/elapid/contributing/) and set up a dev environment with `pre-commit` installed. Then run `pre-commit run --all` to apply formatting.
- To simplify the code, and to better align with other `sklearn` metrics, I'd recommend breaking the `boyce_index` function into smaller components. One to compute the intervals, one for the p/e ratio, one for plotting, and one for returning the correlation coefficient. As a user, I'd expect a single value returned from the boyce_index function.
- I haven't had much time to evaluate the scientific merit of the code yet, so I'll likely provide another review after addressing the key software comments I've made. But what strikes me at the moment is that:
  - The relationships between `nclass`, `window` and `res` are not super clear to me, and the results appear very sensitive to these parameters.
  - It is very easy to return `nan` values, despite the many `nan` checks that are applied. This makes me think that something is not sufficiently robust.

Thanks for your submission, and I'll look forward to your updates.

File: elapid/evaluate.py (hunk 1 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: Done as per suggestion.

File: elapid/evaluate.py (hunk 2 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: Done as per suggestion.

File: elapid/evaluate.py (hunk 3 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: Done as per suggestion.

File: elapid/evaluate.py (hunk 4 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: Done as per suggestion.

File: elapid/evaluate.py (hunk 5 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: removed.

File: elapid/evaluate.py (hunk 7 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: I refactored the code. Intervals are now calculated in a new function, and refactored the code, turned out res was not required at all. Updated the name of the argument to more logical names.

File: elapid/evaluate.py (hunk 8 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: Done

File: elapid/evaluate.py (hunk 9 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: Remove extra nan checking.

File: elapid/evaluate.py (hunk 11 above)
Comment (PC-FSU) [reply elapid/evaluate.py]: So I misinterpreted the equation. For P/E ratio, you need the `predicted frequency` for each habitat suitability region and need only presence-only data.  `Expected frequency` is calculated using the random distribution only. i.e at background points and doesn't required prediction at presence data. I changed the code and docs to reflect the same.

**Comment by PC-FSU** [comment]
Any updates?

File: elapid/evaluate.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+    """
+    Compute the continuous Boyce index to evaluate habitat suitability models.
+
+    Uses the convention as defined in Hirzel et al. 2006 to compute the ratio of observed to expected frequencies.
+
+    Args:
+        yobs (numpy.ndarray | pd.Series | gpd.GeoSeries): Suitability values at observed location (i.e., predictions at presence points).
+        ypred (numpy.ndarray | pd.Series | gpd.GeoSeries): Suitability values at random location (i.e., predictions at background points).
+        nbins (int | list, optional): Number of classes or a list of class thresholds. Defaults to 0.
+        bin_size (float | str, optional): Width of the the bin. Defaults to 'default' which sets width as 1/10th of the fit range.
```
Comment (earth-chris) [inline elapid/evaluate.py]: the interactions between these parameters isn't super clear to me. the default behavior sets `nbins` to 10. setting 0 here doesn't mean that zero bins are estimated, but instead behaves as if no bins are passed at all, which depends on the 'default' parameter.

I might recommend one of the following:

- set the defaults as `nbins=10` and `bin_size=None`, support passing None for each, and throw an error if both are passed (since they are mutually exclusive)
- drop the `bin_size` argument and just go with `nbins`.

File: elapid/evaluate.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+    Returns:
+        np.ndarray: An array of intervals, each represented by a pair of lower and upper bounds.
+
+    Raises:
+        ValueError: If invalid values are provided for nbins or bin_size.
+    """
+    mini, maxi = range
+
+    if isinstance(bin_size, float):
+        nbins = (maxi - mini) / bin_size
```
Comment (earth-chris) [inline elapid/evaluate.py]: I get the following warning that appears to be driven by floating point precision. I don't think it's a problem, just flagging it:

```
ela.evaluate.continuous_boyce_index(y, ypred, bin_size=0.25)
/home/cba/src/elapid/elapid/evaluate.py:59: UserWarning: bin_size has been adjusted to nearest appropriate size using ceil, as range/bin_size : 0.9999999999983606 / 0.25 is not an integer.
```

File: elapid/evaluate.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    if to_plot:
+        plot_PE_curve(x=intervals_mid, y=f_valid)
+
+    results = {
+        "F.ratio": f_scores,
+        "Spearman.cor": corr,
+        "HS": intervals,
+    }
+
+    return results
```
Comment (earth-chris) [inline elapid/evaluate.py]: my personal preference is to avoid returning a dictionary, but instead return the three values as a tuple, so users can specify:

```python
f_ratio, cor, hs = ela.evaluate.continuous_boyce_index(presence, background)
```

File: elapid/evaluate.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+        fi = np.nan  # Avoid division by zero
+    else:
+        fi = pi / ei
+
+    return fi
+
+
+def continuous_boyce_index(
+    yobs: Union[np.ndarray, pd.Series, gpd.GeoSeries],
+    ypred: Union[np.ndarray, pd.Series, gpd.GeoSeries],
```
Comment (earth-chris) [inline elapid/evaluate.py]: I think I had misunderstood the original implementation, and it looks like these are both actually predicted values, just predictions at different locations (ypred at presence and background sites).

I might prefer we rename these variables to be a bit more clear (like `ypred_observed` and `ypred_background`).

**Review by earth-chris — COMMENTED** [review]
Looks much better, @PC-FSU, thank you for the improvements! One more round and I think it's there. There are a few small updates I've requested in code, but I'll make one more here regarding the notebook.

The example provided uses some dummy array data, which does not provide much insight into what the index does in a real-life context. Would you please use modeled predictions from the notebook to demonstrate what it can tell you beyond what you see from the AUC results?

**Comment by earth-chris** [comment]
@PC-FSU let me know if you plan to finish up this contribution. If not, I'll go ahead and merge it and make the remaining updates. Cheers,

**Comment by PC-FSU** [comment]
@earth-chris Please give me some time. I will be able to push it after 2nd April.

**Comment by PC-FSU** [comment]
I've added the suggestions you mentioned. I also spent some time thinking about how to improve the explanation of the Boyce Index in the notebook, but I couldn't come up with a concise way to do it. Could you clarify what you had in mind? The example in the notebook feels fairly limited.

**Comment by earth-chris** [comment]
I've reviewed this PR again and want to document the next steps.

PF-FSU has contributed a large PR with several pieces. There are still a number of changes I want to make to simplify these new features, however. I think this will best be done by merging this PR then submitting the new changes. These include:

- update the CBI call to accept `yobs, ypred` as the inputs instead of having users pre-index the background/presence points
- update the plotting function to compute the CBI values then make the plot, which currently requires x/y values that are the outputs from the CBI function (the mean intervals and f scores, respectively).
- remove some redundant and verbose type checking and instead simply use `.asarray()` to enforce type harmonization
- remove the use of a string 'default' option and instead set a numerical default
- only import the cbi and plotting functions to the main API
- add the evaluate.py module to the docs
- add clearer usage examples in the simple maxent model notebook (instead of the overloaded geospatial data notebook)
